### ISSUE TITLE
[Copy] Duplication of layouts does not retain position and size of all elements.

### ISSUE THREAD

**Issue by akhenry** [nasa/openmctweb#428]
This relates to issue #332 , for which a fix was prepared for the second Baxter testathon. It appears that this fix has not fully resolved the issue. Although some objects retain their position within the layouts, not all do. To reproduce:

1. Duplicate all RP15 Layouts
2. Once duplicated, navigate to Location Layout. RMM_STATE_DAT and LOC_POSE_DAT have retained their position and size within the layout, but other objects have not.

~~More worryingly, in the Thermal Layout, EPSIO_THM_DAT is entirely absent from the layout, although it has been copied successfully and is a child of the Thermal Layout itself.~~ - __edit__: Although this is true, the element is also absent from the copied parent, so is not as a consequence of this issue.

**Comment by VWoeltjen** [comment]
Are any exceptions visible on console when this occurs? Wondering if there are unhandled cases (nulls, perhaps) in the fix for #332

**Comment by akhenry** [comment]
There are no exceptions in the logs other than the 404's on get, which I expect to see due to #378

**Comment by akhenry** [comment]
including an example of one of the 404s in case it is relevant
```
GET http://10.10.81.15:9100/mct/domain_object/0a98586a-99b0-4641-9c64-8083120e492d 404 (Not Found) angular.js:10661
(anonymous function) angular.js:10661
q angular.js:10480
g angular.js:10187
(anonymous function) angular.js:14634
n.$eval angular.js:15916
n.$digest angular.js:15727
(anonymous function) angular.js:15955
e angular.js:5490
(anonymous function) angular.js:5762
```

**Comment by VWoeltjen** [comment]
Trying this locally - seems to work correctly when I just Duplicate Location Layout, but does fail as described when I Duplicate the whole folder of layouts. (This difference may effect how we assess severity of this issue.)

In terms of root cause, think this is may be an ordering problem - the `idMap` from #378 may not be fully populated at the time IDs in the model get rewritten.

**Comment by akhenry** [comment]
Hrm. The replacement of IDs happens in a [discrete step](https://github.com/nasa/openmctweb/blob/0cfc070f3cc5f02a84a969f598cd78e31e266806/platform/entanglement/src/services/CopyTask.js#L248) after the in-memory copy has taken place. All nodes of the tree should have been visited and the idmap fully populated. The exception to this is non-createable types, the algorithm will not visit the children of non-createable types because instead it just creates a link to the parent, whose children are then automatically included by composition. This should not be a problem, we don't need to replace IDs of linked objects as they retain their original ID.

**Comment by akhenry** [comment]
Edited original comment to remove reference to EPSIO_THM_DAT being absent.

**Comment by VWoeltjen** [comment]
Looking at this currently; when the check that [should rewrite the relevant part of Layout configuration](https://github.com/nasa/openmctweb/blob/0cfc070f3cc5f02a84a969f598cd78e31e266806/platform/entanglement/src/services/CopyTask.js#L126) occurs, the ID for the relevant panel is not in the idMap.

**Comment by VWoeltjen** [comment]
Ah! Understand why this is happening: It occurs when an object occurs more than once in a copy operation.

For instance, imagine the following graph (apostrophe indicates a link):

* Folder A
  * Layout A
    * Panel A
  * Layout B
    * Panel A'
  
Links get cloned during a duplicate, so this becomes:

* Folder X
  * Layout X
    * Panel X
  * Layout Y
    * Panel Y

Where there had been one panel (Panel A), there are now two copies (Panel X and Panel Y).

However, the ID map looks like:

Old | New
----|----
Folder A | Folder X
Layout A | Layout X
Layout B | Layout Y
Panel A | Panel Y

So, when Layout X is updated, all references to Panel A are replaced with Panel Y. But Layout X doesn't contain Panel Y, it contains Panel X, and it has no idea how to position it.

I'll look at options for fixing this. I think maintaining a separate idMap per-object might work. There is one hypothetical case where this would still fail (imagine if Folder X had some meaningful reference to Panel A above) but we don't have any practical cases where this occurs. The alternative would be a behavioral change, such that you end up with Panel X and Panel X'; think this would be confusingly inconsistent with other copy rules for links.

**Comment by akhenry** [comment]
Ah, yep, that makes sense. Good analysis. 

Are there cases where the objects will hold references to other objects that they do not compose? The impetus for this change was layout configuration, and in this case the objects are all composed by the parent. Are there other cases where this is not true?

If not, we could use the bottom-up to our advantage. If the scope of the idMap were limited to, say, the copyComposees function, we could do the id replacements at that stage, and it would save us having to keep object-specific id maps. So instead of one big replace at the end, with a global id map, we just keep an id map of the composees of a child, and right after the composees are cloned, we replace their IDs in in the parent.

**Comment by VWoeltjen** [comment]
Per discussion 12/15/2015, doing the per-object ID map is the preferred solution; avoids introducing change in functionality which has not been requested by users

**Comment by akhenry** [comment]
Fixed in PR https://github.com/nasa/openmctweb/pull/435

**Comment by akhenry** [comment]
Reopening due to reappearance in testathon-3.

**Comment by akhenry** [comment]
Error occurs during copy in testathon-3, missing idMap reference. Case is for id replacement in arrays.
https://github.com/nasa/openmctweb/blob/master/platform/entanglement/src/services/CopyTask.js#L120

**Comment by VWoeltjen** [comment]
Remaining issues addressed by #439
